lib/et/com_err.3
.\" Copyright (c) 1988 Massachusetts Institute of Technology,
.\" Student Information Processing Board.  All rights reserved.
.\"
.\" $Header$
.\"
.TH COM_ERR 3 "22 Nov 1988" SIPB
.SH NAME
com_err \- common error display routine
.SH SYNOPSIS
.nf
 #include <com_err.h>
.PP
void initialize_XXXX_error_table(void);
.PP
void com_err(const char *\fIwhoami\fP, long \fIcode\fP,
	const char *\fIformat\fP, ...);
.PP
const char *error_message(long \fIcode\fP);
.PP
const char *error_message_r(long \fIcode\fP, char *\fIbuf\fP);
.PP
typedef void(*error_handler_t)(const char *\fIwhoami\fP,
	long \fIcode\fP, const char *\fIformat\fP, va_list \fIargs\fP);
.PP
error_handler_t set_com_err_hook(error_handler_t \fIproc\fP);
.PP
error_handler_t reset_com_err_hook(void);
.fi
.SH DESCRIPTION
\fICom_err\fP displays an error message on the standard error stream
\fIstderr\fP (see
.IR stdio (3S))
composed of the \fIwhoami\fP string, which should specify the program
name or some subportion of a program, followed by an error message
generated from the \fIcode\fP value (derived from
.IR compile_et (1)),
and a string produced using the \fIformat\fP string and any following
arguments, in the same style as
.IR fprintf (3).

The behavior of \fIcom_err\fP can be modified using
\fIset_com_err_hook\fP; this defines a procedure which is called with
the arguments passed to \fIcom_err\fP, instead of the default internal
procedure which sends the formatted text to error output.  Thus the
error messages from a program can all easily be diverted to another
form of diagnostic logging, such as
.IR syslog (3).
\fIReset_com_err_hook\fP may be used to restore the behavior of
\fIcom_err\fP to its default form.  Both procedures return the
previous ``hook'' value.  These ``hook'' procedures must have the
declaration given for \fIproc\fP above in the synopsis.

The \fIinitialize_XXXX_error_table\fP routine is generated
mechanically by
.IR compile_et (1)
from a source file containing names and associated strings.  Each
table has a name of up to four characters, which is used in place of
the \fBXXXX\fP in the name of the routine.  These routines should be
called before any of the corresponding error codes are used, so that
the \fIcom_err\fP library will recognize error codes from these tables
when they are used.

The functions \fIerror_message\fP and \fIerror_message_r\fP return the
error message for an error code.  \fIerror_message\fP may return its
error message inside a static buffer, and therefore is not reentrant;
\fIerror_message_r\fP will in these cases use the storage provided in
\fIbuf\fP, which should be at least \fBCOM_ERR_BUF_LEN\fP bytes long.

The \fBcom_err.h\fP header file should be included in any source file
that uses routines from the \fIcom_err\fP library; executable files
must be linked using \fI``-lcom_err''\fP in order to cause the
\fIcom_err\fP library to be included.

Multithreaded programs must perform all of the
\fIinitialize_XXXX_error_table\fP calls serially, but may perform them
in parallel with \fIcom_err\fP, \fIerror_message\fP, and
\fIerror_mesesage_r\fP calls as long as the call to
\fIinitialize_XXX_error_table\fP for a given table has completed
before any corresponding error codes are used.  Calls to
\fIset_com_err_hook\fP and \fIreset_com_err_hook\fP must also be
performed serially or they may return the wrong old hook value.

.\" .IR for manual entries
.\" .PP for paragraph breaks

.SH "SEE ALSO"
compile_et (1), syslog (3).

Ken Raeburn, "A Common Error Description Library for UNIX".


ChangeLog
2001-09-23  Ken Raeburn  <raeburn@mit.edu>

	* configure.in: Check for X11/Xaw/Label.h.  If it's not found,
	don't build xzwrite.

	* aclocal.m4 (ATHENA_RESOLV_CHECK): New macro.  Check for
	__res_send as well as res_send.
	(ATHENA_KRB5_CHECK, ATHENA_KRB4_CHECK, ATHENA_HESIOD_CHECK,
	ATHENA_ARES_CHECK): AC_REQUIRE it, so we only include the code
	once.
	(ATHENA_KRB5_CHECK, ATHENA_KRB4_CHECK): Include $LIBS when trying
	to link.
	* acconfig.h (HAVE_LIB_RESOLV): Undef, since acconfig won't cope
	with the explicit definition now used.

2001-08-25  Ken Raeburn  <raeburn@mit.edu>

	* Makefile.in (h/config.h, h/config.h.in): New targets.
	(all): Depend on h/config.h.
	* acconfig.h (HAVE_SA_LEN): New macro.

2001-08-25  Ken Raeburn  <raeburn@mit.edu>

	* Makefile.in (VPATH): Define.
	(configure): Switch to srcdir before running autoconf.

	* configure.in: Check for ifaddrs.h.
	(KRB5_SOCKADDR_SA_LEN): Define and use.  Definition copied from
	krb5 tree.

2001-03-26  Ken Raeburn  <raeburn@mit.edu>

	* Makefile.in (Makefile, config.status, configure): New targets.



lib/et/com_err.c
/*
 * Copyright 1987, 1988 by MIT Student Information Processing Board.
 *
 * For copyright info, see mit-sipb-copyright.h.
 */

#include <stdio.h>
#include <sysdep.h>
#include "mit-sipb-copyright.h"
#include "error_table.h"

#ifndef HAVE_VPRINTF
/* We don't have the v*printf routines... */
#define vfprintf(stream,fmt,args) _doprnt(fmt,args,stream)
#endif

/*
 * Protect us from header version (externally visible) of com_err, so
 * we can survive in a <varargs.h> environment.  I think.
 */
#define com_err com_err_external
#include "com_err.h"
#undef com_err

#if ! lint
static const char rcsid[] =
    "$Header$";
#endif	/* ! lint */

static void
default_com_err_proc (whoami, code, fmt, args)
    const char *whoami;
    long code;
    const char *fmt;
    va_list args;
{
    char buf[25];

    if (whoami) {
	fputs(whoami, stderr);
	fputs(": ", stderr);
    }
    if (code) {
	fputs(error_message_r(code, buf), stderr);
	fputs(" ", stderr);
    }
    if (fmt) {
        vfprintf (stderr, fmt, args);
    }
    /* should do \r only on a tty in raw mode, but it won't hurt */
    putc('\r', stderr);
    putc('\n', stderr);
    fflush(stderr);
}

error_handler_t com_err_hook = default_com_err_proc;

void com_err_va (whoami, code, fmt, args)
    const char *whoami;
    long code;
    const char *fmt;
    va_list args;
{
    (*com_err_hook) (whoami, code, fmt, args);
}

#ifdef __STDC__
void com_err (const char *whoami,
	      long code,
	      const char *fmt, ...)
{
#else
void com_err (whoami, code, fmt, va_alist)
    const char *whoami, *fmt;
    long code;
    va_dcl
{
#endif
    va_list pvar;

    VA_START(pvar, fmt);
    com_err_va (whoami, code, fmt, pvar);
    va_end(pvar);
}

error_handler_t set_com_err_hook (new_proc)
    error_handler_t new_proc;
{
    error_handler_t x = com_err_hook;

    if (new_proc)
	com_err_hook = new_proc;
    else
	com_err_hook = default_com_err_proc;

    return x;
}

error_handler_t reset_com_err_hook () {
    error_handler_t x = com_err_hook;
    com_err_hook = default_com_err_proc;
    return x;
}
